vaadin-grid-flow-parent/vaadin-grid-flow/src/test/java/com/vaadin/flow/component/grid/GridTest.java

```diff
@@ -111,4 +113,60 @@ public void dataProviderListeners_gridAttachedAndDetached_oldDataProviderListene
                         new Grid<>(), 2, 2, new int[] { 0, 2 },
                         new DataCommunicatorTest.MockUI());
     }
+
+    @Test
+    public void setSmallPageSize_callSetRequestedRangeWithLengthLargerThan500_doesNotThrowException() {
+        final Grid<String> grid = new Grid<>();
+
+        grid.setPageSize(10);
+        callSetRequestedRange(grid, 0, 600);
+    }
+
+    @Test
+    public void setAllRowsVisible_setLargePageSize_callSetRequestedRangeWithLengthLargerThan500_doesNotThrowException() {
+        final Grid<String> grid = new Grid<>();
+
+        grid.setPageSize(100);
+        grid.setAllRowsVisible(true);
+        callSetRequestedRange(grid, 0, 600);
+    }
+
+    @Test
+    public void setAllRowsVisible_setSmallPageSize_callSetRequestedRangeWithLengthSmallerThan500_doesNotThrowException() {
+        final Grid<String> grid = new Grid<>();
+
+        grid.setPageSize(10);
+        grid.setAllRowsVisible(true);
+        callSetRequestedRange(grid, 0, 400);
+    }
+
+    @Test
+    public void setAllRowsVisible_setSmallPageSize_callSetRequestedRangeWithLengthLargerThan500_throwsException() {
+        exceptionRule.expect(IllegalArgumentException.class);
+        exceptionRule.expectMessage(
+                "Attempted to fetch more items from server than allowed in one go");
+
+        final Grid<String> grid = new Grid<>();
+
+        grid.setPageSize(10);
+        grid.setAllRowsVisible(true);
+        callSetRequestedRange(grid, 0, 600);
+    }
+
+    private void callSetRequestedRange(Grid<String> grid, int start,
+            int length) {
+        try {
+            Method method = Grid.class.getDeclaredMethod("setRequestedRange",
+                    int.class, int.class);
+            method.setAccessible(true);
+            method.invoke(grid, start, length);
+        } catch (NoSuchMethodException | SecurityException
+                | IllegalAccessException | IllegalArgumentException
+                | InvocationTargetException e) {
+            if (e.getCause() instanceof IllegalArgumentException) {
+                throw (IllegalArgumentException) e.getCause();
+            }
+            Assert.fail("Could not call Grid.setRequestedRange");
```

Comment:
It would be helpful to know the cause of the failure here. Could we simply not handle these exceptions like [here](https://github.com/vaadin/flow-components/pull/3823/files#diff-e54ae4713dd2efeae52550fd49e826c987db76c85171182b6fda8e70c8ba4cc3R20-R28)? Does that approach have any negative consequences?